### PR TITLE
new scc::tlm_target_bfs. A scc::tlm_target based class with callback-enabled bitfields (updated PR)

### PR DESCRIPTION
**Pull request by JoGei**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -169,4 +169,3 @@ add_subdirectory(third_party)
 if(NOT SCC_LIB_ONLY)
 	add_subdirectory(examples)
 endif()
```
Comment (JoGei) [inline CMakeLists.txt:171]: trailing whitespaces in upstream

File: examples/CMakeLists.txt [hunk 2]
```diff
@@ -3,3 +3,4 @@ if(USE_CWR_SYSTEMC)
 	add_subdirectory(axi-axi)
 endif()
 add_subdirectory(axi-pinlevel)
+add_subdirectory(scc-tlm_target_bfs)
```
Comment (JoGei) [inline examples/CMakeLists.txt:6]: add example for tlm_target_bfs module to default build configuration

File: examples/ace-axi/sc_main.cpp [hunk 3]
```diff
@@ -24,6 +24,7 @@
 using namespace sc_core;
 using namespace axi;
 using namespace axi::pe;
+using namespace sc_core;
```
Comment (JoGei) [inline examples/ace-axi/sc_main.cpp:27]: residue from old pull-request https://github.com/VP-Vibes/SystemC-Components/pull/4
might not be needed anymore?

File: examples/scc-tlm_target_bfs/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+project (scc-tlm_target_bfs-example)
+
+add_executable(${PROJECT_NAME} sc_main.cpp)
+
+target_include_directories(${PROJECT_NAME} PRIVATE ${CMAKE_CURRENT_SOURCE_DIR})
+target_link_libraries (${PROJECT_NAME} PUBLIC tlm-interfaces)
+target_link_libraries (${PROJECT_NAME} PUBLIC scc)
+target_link_libraries (${PROJECT_NAME} LINK_PUBLIC ${CMAKE_THREAD_LIBS_INIT})
+target_link_libraries (${PROJECT_NAME} LINK_PUBLIC ${CMAKE_DL_LIBS})
```
Comment (JoGei) [inline examples/scc-tlm_target_bfs/CMakeLists.txt:11]: build configuration for tlm_target_bfs example

File: examples/scc-tlm_target_bfs/sc_main.cpp [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+      /* .num_irqs = */   0,
+      /* .num_regs = */   3
+    };
+    test.addPeripheral<tlm_target_bfs_example>("per1", std::move(per1params));
+
+    sc_core::sc_start(1_ms);
+    SCCINFO() << "Finished";
+
+    return 0;
+}
```
Comment (JoGei) [inline examples/scc-tlm_target_bfs/sc_main.cpp:249]: tlm_target_bfs full example

File: src/components/scc/tlm_target_bfs.h [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    if (rst_in_.read())
+      regs->reset_start();
+    else
+      regs->reset_stop();
+  }
+};
+
+}  // namespace scc
+
+#endif // __SCC_TLM_TARGET_BFS_H__
```
Comment (JoGei) [inline src/components/scc/tlm_target_bfs.h:145]: new scc module tlm_target_bfs

File: src/sysc/scc/ordered_semaphore.h [hunk 7]
```diff
@@ -155,12 +155,13 @@ class SC_API ordered_semaphore : public sc_core::sc_semaphore_if, public sc_core
...
                 }
-            } else {
+            } else if(queue[0].size()) {
                 if(queue[0].front() == sc_core::sc_get_current_process_handle()) {
                     queue[0].pop_front();
                     return false;
                 }
-            }
+            } else
+            	return false;
```
Comment (JoGei) [inline src/sysc/scc/ordered_semaphore.h:164]: residue from old pull-request https://github.com/VP-Vibes/SystemC-Components/pull/4
Not sure where this is from, or what it does. @eyck ?

**Review by JoGei — COMMENTED** [review]
Needs some work.

File: src/sysc/scc/ordered_semaphore.h [hunk 8]
```diff
@@ -155,12 +155,13 @@ class SC_API ordered_semaphore : public sc_core::sc_semaphore_if, public sc_core
...
                     return false;
                 }
-            } else {
+            } else if(queue[0].size()) {
                 if(queue[0].front() == sc_core::sc_get_current_process_handle()) {
                     queue[0].pop_front();
                     return false;
                 }
-            }
+            } else
```
Comment (eyck) [inline src/sysc/scc/ordered_semaphore.h:163]: These two lines stem from an earlier (and buggy version). If non of the queues (which set-up 2 priorities) have entries the in_use() function should return true.

**Review by eyck — APPROVED** [review]

**Pull request merged**